### PR TITLE
Feature/#1799 full screen screenshots

### PR DESCRIPTION
**Pull request by asolntsev**
fix for #1799, continuation of https://github.com/selenide/selenide/pull/1800

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/impl/WebdriverPhotographer.java [hunk 1]
```diff
@@ -1,24 +1,94 @@
...
 
 @ParametersAreNonnullByDefault
 public class WebdriverPhotographer implements Photographer {
+  private static final JavaScript js = new JavaScript("get-screen-size.js");
+
   @Nonnull
   @CheckReturnValue
   @Override
   public <T> Optional<T> takeScreenshot(Driver driver, OutputType<T> outputType) {
+    if (driver.getWebDriver() instanceof ChromiumDriver chromiumDriver) {
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/WebdriverPhotographer.java:31]: You cannot rely on such type guard. Webdriver instance can be decorated e.g. from webdriver listeners.
example:
```
public class MainPageTest {
    @Test
    public void test() {
        WebDriverRunner.addListener(new WebDriverListener() {
            @Override
            public void beforeAnyCall(Object target, Method method, Object[] args) {

            }
        });
        Selenide.open();
        WebDriver driver = WebDriverRunner.getWebDriver();
        System.out.println(driver.getClass()); // net.bytebuddy.renamed.java.lang.Object$ByteBuddy$gM4k6yJv
        System.out.println(Arrays.toString(driver.getClass().getInterfaces())); // all ChromeDriver interfaces
        assertTrue(driver instanceof ChromiumDriver); // fail :/
    }
}
```

**Comment by asolntsev** [comment]
@rosolko I measured how slow the screenshooting is.

Generating 100 screenshoots took:

|       | Chrome | Chrome headless | Firefox |
| ----------- | ----------- | ----------- | ----------- |
| Old good screenshots      | 30s   |  22s       | 15s |
| Full-screen screenshots   | 35s   |  27s       | 3min |

In Chrome:
* The full-screen screenshots are a bit slower, but the difference is minor. 

In Firefox:
* **The full-screen screenshots are a much slower!**

File: statics/src/test/java/integration/ScreenshotsTest.java [hunk 2]
```diff
@@ -52,4 +52,15 @@ void canTakeScreenshotAsTemporaryFile() throws IOException {
       .hasSize(1);
     assertThat(screenshots.getContextScreenshots().get(0)).isSameAs(screenshot);
   }
+
+  @Test
+  void canTakeFullScreenshot() throws IOException {
```
Comment (BorisOsipov) [inline statics/src/test/java/integration/ScreenshotsTest.java:57]: What this test supposed to test? I changed `WebdriverPhotographer` to old default implementation without full screens support and test won't fail on my pc. I supposed it should fail if we don't take full screen screenshot. What am I missing?
image

Comment (asolntsev) [reply statics/src/test/java/integration/ScreenshotsTest.java:57]: Yes, these checks for size a crap. Need to review them.

File: src/main/java/com/codeborne/selenide/impl/WebdriverPhotographer.java (hunk 1 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/impl/WebdriverPhotographer.java]: @BorisOsipov Valid point. Fixed in https://github.com/selenide/selenide/pull/1858.

File: statics/src/test/java/integration/ScreenshotsTest.java (hunk 2 above)
Comment (asolntsev) [reply statics/src/test/java/integration/ScreenshotsTest.java:57]: @BorisOsipov Valid point. Fixed in https://github.com/selenide/selenide/pull/1858 (old tests check the old logic, new tests check the new logic).

**Comment by asolntsev** [comment]
reworked in https://github.com/selenide/selenide/pull/1858
(extracted full-size screenshooting to a separate Selenide plugin).

**Pull request closed without merging**